File: 0061-rotate-list/0061-rotate-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode rotateRight(ListNode head, int k) {
        ListNode dummy = new ListNode(), slow, fast;
        dummy.next = slow = head;

        int size = 0;
        while(slow != null) {
            size++;
            slow = slow.next;
        }

        slow = fast = dummy;
        int corr = size != 0 ? (k % size) : 0;

        while(fast.next != null) {
            if(corr > 0) {
                fast = fast.next;
                corr--;
            } else {
                slow = slow.next;
                fast = fast.next;
            }
        }

        if(slow != fast) {
            fast.next = head;
            dummy.next = slow.next;
            slow.next = null;
        }

        return dummy.next;
    }
}

File: 0061-rotate-list/0061-rotate-list.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* rotateRight(struct ListNode* head, int k) {
    struct ListNode* dummy = malloc(sizeof(struct ListNode)), *temp, *slow, *fast;
    slow = fast = dummy;
    dummy -> next = temp = head;
    int size = 0;

    while(temp != NULL) {
        size++;
        temp = temp -> next;
    }

    int corr = size ? (k % size) : 0;

    while(corr--) fast = fast -> next;

    while(fast -> next != NULL) {
        slow = slow -> next;
        fast = fast -> next;
    }

    if(slow != fast) {
        fast -> next = head;
        dummy -> next = slow -> next;
        slow -> next = NULL;
    }

    return dummy -> next;
}